src/corelibs/U2Gui/src/util/U2FileDialog.cpp

```diff
@@ -94,11 +108,68 @@ QString U2FileDialog::getSaveFileName(QWidget* parent,
                                       const QString& caption,
                                       const QString& dir,
                                       const QString& filter,
-                                      QString* selectedFilter,
+                                      const QString& selectedFilter,
                                       const QFileDialog::Options& options) {
     activateAppWindow();
     QFileDialog::Options effectiveOptions = getEffectiveOptions(options);
-    return QFileDialog::getSaveFileName(parent, caption, dir, filter, selectedFilter, effectiveOptions);
+    return U2FileDialog::getFileName(parent,
+                                     caption,
+                                     dir,
+                                     filter,
+                                     selectedFilter,
+                                     effectiveOptions,
+                                     QFileDialog::AcceptSave,
+                                     QFileDialog::ExistingFile);
 }
 
+QStringList U2FileDialog::getFileNames(QWidget* parent,
+                                       const QString& caption,
+                                       const QString& dir,
+                                       const QString& filter,
+                                       const QString& selectedFilter,
+                                       QFileDialog::Options options,
+                                       QFileDialog::AcceptMode acceptMode,
+                                       QFileDialog::FileMode fileMode) {
+
+    QFileDialog fileDialog(parent, caption, dir, filter);
+    if (!selectedFilter.isEmpty()) {
+        fileDialog.selectNameFilter(selectedFilter);
+    }
+    fileDialog.setOptions(options);
+    fileDialog.setFileMode(fileMode);
+    fileDialog.setAcceptMode(acceptMode);
+    auto deleteConnection = QObject::connect(parent, &QWidget::destroyed, [&]() {
```

Comment:
So we want to allow deletion of parent of QFileDialog, right? 
In this case we should follow the following logic: when a parent of a widget is deleted the widget is deleted too.

Here you keep the dialog alive for a longer period, than the parent: the parent is deleted, but the dialog stays alive.

This does not look a correct behavior and has side-effects: close() will cause CloseEvent to be sent to all parties first, next the dialog will process CloseEvent and send signals (rejected/done) to all parties and only next it will be destroyed. This all happens when the application hierarchy is already partially destroyed: destroyed() is emitted from inside of the main window object destructor.  I would follow a simple rule: never listen to 'destroyed' within the same child<->parent hierarchy.

I see 2 simpler solutions to check:

1) We can disable file dialog invocation during UGENE shutdown completely: either by disabling the related actions or/and, by creating a special flag that shutdown is in progress and using this flag inside U2FileDialog factory methods: the dialog won't be shown at all if UGENE is closing (CHECK(!isShuttingDown)).

2) Allow destroy of the dialog with the parent. Use QPointer<> to track if the dialog was destroyed inside of the U2FileDialog methods and return empty result if it was.

Reply:
I don't see a point into a not-trusting the `destroyed` signal, which is called right before the object is deleted and supposed to be used just for the cases like this. This is a Qt very-very base architecture, if we cannot trust it we cannot trust anything in our code at all. That means, that if we have a rule "never listen to 'destroyed'", than we should have a rule "never listen to anything which is produced by Qt".
And we cannot - just really cannot - delete the dialog when `exec()` is still working - it will definitely make a crash. Where is nothing bad in **existing** of the `QFileDialog fileDialog` object, it doesn't creates any problems. The problem appears because it's **executing** when parent dyes, not because it just **exists**.
So, my solution looks like it uses the exact functions which should be used in this exact case (actually, I spied it somewhere on stackoverflow, so this is a pretty well-known solution), I see what you said, but see no point of changing it.

Reply:
What will happen it you allow deletion of the dialog together with the parent (done automatically)? 

If it works (like it works with any other dialog we have in UGENE) we can throw away the 1)subscribe 2) reparent 3) manual close 4) unsubscibe parts in your code.

Reply:
You cannot delete parent while the dialog is executing, this fact is even documented:
![image](https://user-images.githubusercontent.com/26257527/187844227-ffcd6e82-d948-4359-bd45-0a290269b47b.png)

Reply:
Right, and you already do what is written here: you create the dialog using QFileDialog constructor. 
Now it is safe to have it auto-deleted with parent.

Reply:
Well, as far as I said - we do not need it, its just some complication you've thought up for I don't know why reason (sorry, but that'' true). The **existing** of the `fileDialog` object doesn't make any problems, its **execution** makes and I handled the execution. The dialog will be destroyed on itself on `getFileNames` closing. It's not a pointer, it's just an object, we do not need to do anything about it, it will be handled on its own when it pops out of the call stack, this is how the C++ language works, I do not know how to explain it anymore.

Let's do this - please, gimme a scenario where we still have a crash of `SAFE_POINT` or whatever and I will fix it.

Reply:
Here is a thing that is better than an example: a complete code snippet.

```c++
    QPointer<QFileDialog> fileDialog(new QFileDialog(parent, caption, dir, filter));
    if (!selectedFilter.isEmpty()) {
        fileDialog->selectNameFilter(selectedFilter);
    }
    fileDialog->setOptions(options);
    fileDialog->setFileMode(fileMode);
    fileDialog->setAcceptMode(acceptMode);

    QStringList result;
    if (fileDialog->exec() == QFileDialog::Accepted && !fileDialog.isNull()) {
        result = fileDialog->selectedFiles();
    }
    return result;
```

We use this pattern across UGENE code everywhere we handle dialogs and QT docs for QFileDialog says this approach is legit for file dialogs too.

Reply:
I see my fix does not work on 5.12.12 on Windows but works on 5.15.2
Lets use your fix in this case.
Please fix other issues (reference types) before submitting.